File: 0079-word-search/0079-word-search.cpp
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    bool rec(vector<vector<char>>& board, string word, int idx, int i, int j) {
        // Base case for recursion
        if (idx == word.length()) {
            return true; // Word found
        }

        // Boundary check
        if (i < 0 || j < 0 || i >= board.size() || j >= board[0].size() || board[i][j] != word[idx]) {
            return false; // Invalid position or mismatched character
        }

        // Mark current cell as visited
        char temp = board[i][j];
        board[i][j] = '#';

        // Explore neighbors
        bool found = rec(board, word, idx + 1, i - 1, j) ||
                     rec(board, word, idx + 1, i + 1, j) ||
                     rec(board, word, idx + 1, i, j - 1) ||
                     rec(board, word, idx + 1, i, j + 1);

        // Restore original state
        board[i][j] = temp;

        return found;
    }

    bool exist(vector<vector<char>>& board, string word) {
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (rec(board, word, 0, i, j)) {
                    return true;
                }
            }
        }
        return false;
    }
};

File: 0079-word-search/0079-word-search.java
class Solution {
    public boolean wordFormed(char[][] board, String word,int idx,int i,int j,boolean[][] visited)
    {
        if(idx==word.length())
        return true;

        if(i<0 || j<0 ||i>=board.length || j>=board[0].length || idx>word.length() || board[i][j]!=word.charAt(idx)||visited[i][j])
        return false;

        visited[i][j]=true;

        boolean found = wordFormed(board, word, idx + 1, i + 1, j, visited) ||
                wordFormed(board, word, idx + 1, i, j + 1, visited) ||
                wordFormed(board, word, idx + 1, i - 1, j, visited) ||
                wordFormed(board, word, idx + 1, i, j - 1, visited);


        visited[i][j]=false;

        return found;
    }

    public boolean exist(char[][] board, String word) {
        int n=board.length;
        int m=board[0].length;
        int len=word.length();
        boolean[][] visited=new boolean[n][m];
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(board[i][j]==word.charAt(0) && wordFormed(board,word,0,i,j,visited))
                return true;
            }
        }

        return false;
    }
}